File: dataset/ourMethod/OPENAI_gpt-5-2025-08-07/Task125.java
import java.util.*;

public class Task125 {

    private static final String GENERIC_ERROR = "ERROR: Access denied or resource not found.";
    private static final Set<String> ALLOWED_ROLES;
    private static final Map<String, Resource> RESOURCES;

    static {
        Set<String> roles = new HashSet<>(Arrays.asList("guest", "user", "admin"));
        ALLOWED_ROLES = Collections.unmodifiableSet(roles);

        Map<String, Resource> res = new HashMap<>();
        res.put("public", new Resource("Welcome to the public area.", new HashSet<>(Arrays.asList("guest", "user", "admin"))));
        res.put("dashboard", new Resource("Admin control panel.", new HashSet<>(Arrays.asList("admin"))));
        res.put("profile", new Resource("User profile page.", new HashSet<>(Arrays.asList("user", "admin"))));
        res.put("report2025", new Resource("Confidential report 2025.", new HashSet<>(Arrays.asList("user", "admin"))));
        res.put("help", new Resource("Help and FAQs.", new HashSet<>(Arrays.asList("guest", "user", "admin"))));
        RESOURCES = Collections.unmodifiableMap(res);
    }

    private static final class Resource {
        private final String content;
        private final Set<String> allowedRoles;

        private Resource(String content, Set<String> allowedRoles) {
            this.content = content;
            this.allowedRoles = Collections.unmodifiableSet(new HashSet<>(allowedRoles));
        }
    }

    public static String accessResource(String role, String resourceId) {
        if (!isValidRole(role) || !isValidResourceId(resourceId)) {
            return GENERIC_ERROR;
        }
        Resource r = RESOURCES.get(resourceId);
        if (r == null) {
            return GENERIC_ERROR;
        }
        if (!r.allowedRoles.contains(role)) {
            return GENERIC_ERROR;
        }
        return r.content;
    }

    private static boolean isValidRole(String role) {
        if (role == null) return false;
        return ALLOWED_ROLES.contains(role);
    }

    private static boolean isValidResourceId(String id) {
        if (id == null) return false;
        int len = id.length();
        if (len < 1 || len > 20) return false;
        for (int i = 0; i < len; i++) {
            char c = id.charAt(i);
            boolean ok = (c >= 'A' && c <= 'Z') ||
                         (c >= 'a' && c <= 'z') ||
                         (c >= '0' && c <= '9') ||
                         c == '_' || c == '-';
            if (!ok) return false;
        }
        return true;
    }

    public static void main(String[] args) {
        // 5 test cases
        String[] roles = {"admin", "guest", "user", "guest", "user"};
        String[] resources = {"dashboard", "public", "report2025", "report2025", "bad$id"};
        for (int i = 0; i < roles.length; i++) {
            String result = accessResource(roles[i], resources[i]);
            System.out.println("Test " + (i + 1) + " (" + roles[i] + ", " + resources[i] + "): " + result);
        }
    }
}

File: dataset/ZeroShot/GEMINI_gemini-2.5-pro/Task125.cpp
#include <iostream>
#include <string>
#include <map>
#include <set>

// In C++, the concept of a main class like in Java isn't standard.
// We use a namespace or simply define functions and use a main() function.
// For consistency with the instructions, we can simulate it with a struct/class or namespace.
// We'll call it Task125.

class Task125 {
private:
    // Define a whitelist of available resources
    static const std::map<std::string, std::string> resources;
    // Define role-based access control permissions
    static const std::map<std::string, std::set<std::string>> rolePermissions;

public:
    /**
     * Securely accesses a resource based on a user's role and requested resource ID.
     *
     * @param resourceId The identifier for the resource to be accessed.
     * @param userRole   The role of the user requesting access.
     * @return A string indicating the result of the access attempt.
     */
    static std::string accessResource(const std::string& resourceId, const std::string& userRole) {
        // 1. Validate user role
        auto roleIt = rolePermissions.find(userRole);
        if (roleIt == rolePermissions.end()) {
            return "Error: Invalid user role provided.";
        }

        // 2. Validate resource ID against the whitelist
        auto resourceIt = resources.find(resourceId);
        if (resourceIt == resources.end()) {
            return "Access Denied: Resource not found.";
        }

        // 3. Check if the user's role has permission for the resource
        const std::set<std::string>& allowedResources = roleIt->second;
        if (allowedResources.count(resourceId)) {
            return "Access Granted. Now accessing: " + resourceIt->second;
        } else {
            return "Access Denied: You do not have permission to access this resource.";
        }
    }
};

// Initialize static members
const std::map<std::string, std::string> Task125::resources = {
    {"1", "Public Document"},
    {"2", "User Profile"},
    {"3", "Admin Panel"},
    {"4", "System Logs"}
};

const std::map<std::string, std::set<std::string>> Task125::rolePermissions = {
    {"GUEST", {"1"}},
    {"USER", {"1", "2"}},
    {"ADMIN", {"1", "2", "3", "4"}}
};


int main() {
    std::cout << "Running test cases..." << std::endl;

    // Test Case 1: Guest accessing a public resource (should be granted)
    std::cout << "Test 1 (GUEST, Resource '1'): " << Task125::accessResource("1", "GUEST") << std::endl;

    // Test Case 2: Guest trying to access an admin resource (should be denied)
    std::cout << "Test 2 (GUEST, Resource '3'): " << Task125::accessResource("3", "GUEST") << std::endl;

    // Test Case 3: Admin accessing an admin resource (should be granted)
    std::cout << "Test 3 (ADMIN, Resource '3'): " << Task125::accessResource("3", "ADMIN") << std::endl;

    // Test Case 4: User trying to access a non-existent resource (should be denied)
    std::cout << "Test 4 (USER, Resource '99'): " << Task125::accessResource("99", "USER") << std::endl;

    // Test Case 5: User accessing their profile (should be granted)
    std::cout << "Test 5 (USER, Resource '2'): " << Task125::accessResource("2", "USER") << std::endl;

    return 0;
}